### PR TITLE
Diffedit Zero-Shot Inpainting Pipeline

### PR DESCRIPTION
**Pull request by clarencechen**
@sayakpaul @patrickvonplaten @silvererudite PR for adding DiffEdit Pipeline introduced in #2800

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 1]
```diff
@@ -750,23 +750,18 @@ def get_epsilon(self, model_output: torch.Tensor, sample: torch.Tensor, timestep
...
+            for j in range(hidden_states.shape[1]):
+                noise = hidden_states[i : i + 1, j : j + 1, :, :]
+                while True:
+                    roll_amount = torch.randint(noise.shape[2] // 2, (1,), generator=generator).item()
+                    reg_loss += (noise * torch.roll(noise, shifts=roll_amount, dims=2)).mean() ** 2
+                    reg_loss += (noise * torch.roll(noise, shifts=roll_amount, dims=3)).mean() ** 2
+
+                    if noise.shape[2] <= 8:
+                        break
+                    noise = F.avg_pool2d(noise, kernel_size=2)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:764]: Let's keep the unrelated files untouched. We can do this in a follow-up PR :)

**Comment by sayakpaul** [comment]
@clarencechen thank you so much for opening the PR. 

Let us know when you would like us to give a review. I noticed you were in between commits; hence was asking.

File: docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# Zero-shot Diffusion-based Semantic Image Editing with Mask Guidance
+
+## Overview
+
+[DiffEdit: Diffusion-based semantic image editing with mask guidance](https://arxiv.org/abs/2210.11427).
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx:17]: Can we also add the authors?

File: docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+Resources:
+
+* [Project Page](https://pix2pixzero.github.io/).
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx:25]: Needs to be edited?

File: docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+this in the pipeline, you simply have to set the embeddings related to the phrases including "cat" to
+`source_embeds` and "dog" to `target_embeds`. Refer to the code example below for more details.
+* When you're using this pipeline from a prompt, specify the _source_ concept in the prompt. Taking
+the above example, a valid input prompt would be: "a high resolution painting of a **cat** in the style of van gough".
+* If you wanted to reverse the direction in the example above, i.e., "dog -> cat", then it's recommended to:
+    * Swap the `source_embeds` and `target_embeds`.
+    * Change the input prompt to include "dog".  
+* To learn more about how the source and target embeddings are generated, refer to the [original 
+paper](https://arxiv.org/abs/2302.03027). Below, we also provide some directions on how to generate the embeddings.
+* Note that the quality of the outputs generated with this pipeline is dependent on how good the `source_embeds` and `target_embeds` are. Please, refer to [this discussion](#generating-source-and-target-embeddings) for some suggestions on the topic.
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx:45]: Needs to be changed?

**Comment by sayakpaul** [comment]
@clarencechen could you please update the PR or the docs with code examples so that we can try them out?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py [hunk 5]
```diff
@@ -0,0 +1,1469 @@
...
+            List of denoised PIL images of length `num_timesteps * batch_size` or numpy array of shape `(num_timesteps,
+            batch_size, height, width, num_channels)`. PIL images or numpy array present the denoised images of the
+            diffusion pipeline.
+    """
+
+    latents: torch.FloatTensor
+    images: Union[List[PIL.Image.Image], np.ndarray]
+
+
+def auto_corr_loss(hidden_states, generator=None):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:63]: Noting for other reviewers. 

It's different from

https://github.com/huggingface/diffusers/blob/d4f846fa74cb713509bab6e3a24d4333eb38a080/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py#L752

Specifically, this implementation supports batches.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py [hunk 6]
```diff
@@ -0,0 +1,1469 @@
...
+        ```py
+        >>> import PIL
+        >>> import requests
+        >>> import torch
+        >>> from io import BytesIO
+
+        >>> from diffusers import StableDiffusionDiffEditPipeline
+
+
+        >>> def download_image(url):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:1049]: Nit. 

We can use `load_image()` from `diffusers.utils` to load an image from a URL. Using this method can save us a couple of LOC.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py [hunk 7]
```diff
@@ -0,0 +1,1469 @@
...
+            if `return_dict` is `True`, otherwise a `tuple`. When returning a tuple, the first element is the inverted
+            latents tensors ordered by increasing noise, and then second is the corresponding decoded images if
+            `decode_latents` is `True`, otherwise `None`.
+        """
+
+        # 1. Check inputs
+        self.check_inputs(
+            prompt,
+            0,
+            0,
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_diffedit.py:1083]: What do these stand for?

**Review by sayakpaul — COMMENTED** [review]
I did a quick pass and overall it looks good to me 🔥 Well done :) 

Thank you for doing this so quickly. 

I think we're only missing out on the tests and other minor edits needed to merge this. Let us know how we can help.

**Comment by clarencechen** [comment]
@sayakpaul I was busy debugging code today, and the current documentation is still a WIP originally copied from Pix2PixZero. I'll push some more updates over the next few days.

**Comment by sayakpaul** [comment]
Sure. Appreciate your willingness :)

**Comment by clarencechen** [comment]
@sayakpaul @patrickvonplaten Opened a PR to add images for pipeline tests at https://huggingface.co/datasets/hf-internal-testing/diffusers-images/discussions/11 - will update to correct directory structure.

**Comment by clarencechen** [comment]
@sayakpaul @patrickvonplaten I know that the two of you have been occupied with other matters lately, but I look forward to getting this new pipeline merged into main soon. Let me know if there are any more outstanding tasks I need to complete or address.

File: docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+
+Resources:
+
+* [Paper](https://arxiv.org/abs/2210.11427).
+* [Blog Post with Demo](https://blog.problemsolversguild.com/technical/research/2022/11/02/DiffEdit-Implementation.html).
+* [Implementation on Github](https://github.com/Xiang-cd/DiffEdit-stable-diffusion/).
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/diffedit.mdx:27]:
```suggestion
* [Original Repository](https://github.com/Xiang-cd/DiffEdit-stable-diffusion/).
```

**Review by sayakpaul — APPROVED** [review]
Pretty cool work! Well done 🔥 

I love the documentation! 

My last points are:

* Could we discard the changes introduced to `src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py`?
* Add an entry about the pipeline to `docs/source/en/api/pipelines/stable_diffusion/overview.mdx`.
* Ensure that `RUN_SLOW=1 pytest tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py` pass?

**Comment by clarencechen** [comment]
> Pretty cool work! Well done 🔥
> 
> I love the documentation!
> 
> My last points are:
> 
> * Could we discard the changes introduced to `src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py`?
> * Add an entry about the pipeline to `docs/source/en/api/pipelines/stable_diffusion/overview.mdx`.
> * Ensure that `RUN_SLOW=1 pytest tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py` pass?

Please merge https://huggingface.co/datasets/hf-internal-testing/diffusers-images/discussions/11 to ensure that the slow tests will pass.

**Comment by sayakpaul** [comment]
> Please merge https://huggingface.co/datasets/hf-internal-testing/diffusers-images/discussions/11 to ensure that the slow tests will pass.

Left a comment there.

**Comment by clarencechen** [comment]
@sayakpaul Unfortunately, I need to change the expected result for the end-to-end pipeline slow test due to some discrepancies with the random generator:

https://huggingface.co/datasets/hf-internal-testing/diffusers-images/discussions/14

@patrickvonplaten I would highly appreciate a final review from you also, as I need to merge this PR for a project I am currently working on.

**Comment by clarencechen** [comment]
@patrickvonplaten @sayakpaul The slow tests should pass now after enforcing determinism in the latent noise for generating the masks, and widening the test tolerance on the model of `test_stable_diffusion_upscale
_pipeline_fp16`, as the slow test also uses fp16.

**Comment by sayakpaul** [comment]
Hi @clarencechen. Thanks for iterating so much here!

I have the following env setup:

```bash
- `diffusers` version: 0.15.0.dev0
- Platform: Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.10
- Python version: 3.8.16
- PyTorch version (GPU?): 1.13.1+cu116 (True)
- Huggingface_hub version: 0.13.2
- Transformers version: 4.27.2
- Accelerate version: 0.18.0
- xFormers version: 0.0.16
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No
```

With this, the tests are failing:

```
============================== short test summary info ==============================
FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_attention_slicing_forward_pass - AssertionError: 0.0027891994 not less than 0.001 : Attention slicing should not ...
FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_inference_batch_single_identical - assert 0.0037350059 < 0.0001
FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_xformers_attention_forwardGenerator_pass - AssertionError: 0.00304389 not less than 0.0001 : XFormers attention should not ...
FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineIntegrationTests::test_stable_diffusion_diffedit_full - torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 63.28 GiB (GP...
=============== 4 failed, 15 passed, 26 warnings in 88.03s (0:01:28) ================
```

Am I missing out on something?

**Comment by clarencechen** [comment]
> test_stable_diffusion_diffedit



> 

The results you've shown look very strange to me. Did you verify that the tests currently in the main branch pass in the same environment?

**Comment by sayakpaul** [comment]
Weird. Not sure what is happening. 

I reinstalled `diffusers` from source:

```
pip install git+https://github.com/huggingface/diffusers
```

Other dependencies are same as mentioned in https://github.com/huggingface/diffusers/pull/2837#issuecomment-1514146801

But still, when I run:

```
RUN_SLOW=1 pytest tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py
```

Could you maybe shed a bit more lights on the dev env you're using to test?

**Comment by clarencechen** [comment]
> Hi @clarencechen. Thanks for iterating so much here!
> 
> I have the following env setup:
> 
> ```shell
> - `diffusers` version: 0.15.0.dev0
> - Platform: Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.10
> - Python version: 3.8.16
> - PyTorch version (GPU?): 1.13.1+cu116 (True)
> - Huggingface_hub version: 0.13.2
> - Transformers version: 4.27.2
> - Accelerate version: 0.18.0
> - xFormers version: 0.0.16
> - Using GPU in script?: Yes
> - Using distributed or parallel set-up in script?: No
> ```
> 
> With this, the tests are failing:
> 
> ```
> ============================== short test summary info ==============================
> FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_attention_slicing_forward_pass - AssertionError: 0.0027891994 not less than 0.001 : Attention slicing should not ...
> FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_inference_batch_single_identical - assert 0.0037350059 < 0.0001
> FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_xformers_attention_forwardGenerator_pass - AssertionError: 0.00304389 not less than 0.0001 : XFormers attention should not ...
> FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineIntegrationTests::test_stable_diffusion_diffedit_full - torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 63.28 GiB (GP...
> =============== 4 failed, 15 passed, 26 warnings in 88.03s (0:01:28) ================
> ```
> 
> Am I missing out on something?

Here is the environment I'm running on:

> ```shell
> - `diffusers` version: 0.16.0.dev0
> - Platform: Linux-5.15.0-52-generic-x86_64
> - Python version: 3.8.10
> - PyTorch version (GPU?): 2.0.0+cu117 (True)
> - Huggingface_hub version: 0.13.4
> - Transformers version: 4.28.1
> - Accelerate version: 0.18.0
> - xFormers version: 0.0.18
> - Using GPU in script?: Yes
> - Using distributed or parallel set-up in script?: No
> ```

It looks like old versions of diffusers dependencies may be causing the OOM issue you see in the slow test. I would really appreciate it if you can assist me with further investigation, but I understand if you can't take on this favor for me.

**Comment by sayakpaul** [comment]
I can check it for you. But let me know when we can lock the commits, as I see there are still commits related to bug fixes (which is fine).

Seeking an opinion from @patrickvonplaten for https://github.com/huggingface/diffusers/pull/2837#issuecomment-1518896146 as well.

**Comment by patrickvonplaten** [comment]
> > Hi @clarencechen. Thanks for iterating so much here!
> > I have the following env setup:
> > ```shell
> > - `diffusers` version: 0.15.0.dev0
> > - Platform: Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.10
> > - Python version: 3.8.16
> > - PyTorch version (GPU?): 1.13.1+cu116 (True)
> > - Huggingface_hub version: 0.13.2
> > - Transformers version: 4.27.2
> > - Accelerate version: 0.18.0
> > - xFormers version: 0.0.16
> > - Using GPU in script?: Yes
> > - Using distributed or parallel set-up in script?: No
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > With this, the tests are failing:
> > ```
> > ============================== short test summary info ==============================
> > FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_attention_slicing_forward_pass - AssertionError: 0.0027891994 not less than 0.001 : Attention slicing should not ...
> > FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_inference_batch_single_identical - assert 0.0037350059 < 0.0001
> > FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineFastTests::test_xformers_attention_forwardGenerator_pass - AssertionError: 0.00304389 not less than 0.0001 : XFormers attention should not ...
> > FAILED tests/pipelines/stable_diffusion_2/test_stable_diffusion_diffedit.py::StableDiffusionDiffEditPipelineIntegrationTests::test_stable_diffusion_diffedit_full - torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 63.28 GiB (GP...
> > =============== 4 failed, 15 passed, 26 warnings in 88.03s (0:01:28) ================
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Am I missing out on something?
> 
> Here is the environment I'm running on:
> 
> > ```shell
> > - `diffusers` version: 0.16.0.dev0
> > - Platform: Linux-5.15.0-52-generic-x86_64
> > - Python version: 3.8.10
> > - PyTorch version (GPU?): 2.0.0+cu117 (True)
> > - Huggingface_hub version: 0.13.4
> > - Transformers version: 4.28.1
> > - Accelerate version: 0.18.0
> > - xFormers version: 0.0.18
> > - Using GPU in script?: Yes
> > - Using distributed or parallel set-up in script?: No
> > ```
> 
> It looks like old versions of diffusers dependencies may be causing the OOM issue you see in the slow test. I would really appreciate it if you can assist me with further investigation, but I understand if you can't take on this favor for me.

Those test failures are expected - that's totally fine! Different hardware give different precisions here

**Review by patrickvonplaten — APPROVED** [review]
This looks good to me overall! Happy to merge actually, @sayakpaul wdyt? 

I don't think the pipeline doesn't work in PyTorch < 2.0 - it's just that in this case it won't make sure of memory-efficient attention by default and requires more memory causing it to fail

**Comment by sayakpaul** [comment]
Works for me then!

**Comment by sayakpaul** [comment]
@clarencechen I want to wholeheartedly thank you for patiently addressing our comments and being so open about incorporating our feedback. Thanks a ton!

**Pull request merged**